src/constructUrl.js

```diff
@@ -113,49 +114,176 @@ var DEFAULT_OPTIONS = Object.freeze({
 
-  return constructUrl(
-    rawSrc,
-    Object.assign(
-      {},
-      params,
-      imgixParams,
-      disableLibraryParam ? {} : { ixlib: `react-${PACKAGE_VERSION}` }
-    )
-  );
+  return constructUrl(rawSrc, {
+    ...params,
+    ...imgixParams,
+    ...(disableLibraryParam ? {} : { ixlib: `react-${PACKAGE_VERSION}` }),
+  });
+}
+
+/**
+ * Build a imgix source url with parameters from a raw url
+ */
+function buildSrc({
+  src: inputSrc,
+  width,
+  height,
+  disableLibraryParam,
+  disableSrcSet,
+  imgixParams,
+  disableQualityByDPR,
+}) {
+  const [rawSrc, params] = extractQueryParams(inputSrc);
+  const fixedSize = width != null || height != null;
+
+  const srcOptions = {
+    ...params,
+    ...imgixParams,
+    ...(disableLibraryParam ? {} : { ixlib: `react-${PACKAGE_VERSION}` }),
+    ...(fixedSize && height ? { height } : {}),
+    ...(fixedSize && width ? { width } : {}),
+  };
+
+  if (disableSrcSet) {
+    const src = constructUrl(rawSrc, srcOptions);
+    return { src, src };
+  }
+
+  if (fixedSize) {
+    const srcSet = buildSrcSet(
+      rawSrc,
+      srcOptions,
+      {
+        disableVariableQuality: disableQualityByDPR,
+      },
+      width,
+      height
+    );
+
+    const src = constructUrl(rawSrc, srcOptions);
+
+    return { src, srcSet };
+  } else {
+    warnInvalidAspectRatio(imgixParams.ar || "", config);
+
+    const { w, h, ...urlParams } = srcOptions;
+    const src = constructUrl(rawSrc, srcOptions);
+    const srcSet = buildSrcSet(rawSrc, { ...urlParams });
+
+    return { src, srcSet };
+  }
+}
+
+function buildSrcSet(rawSrc, params = {}, options = {}, width, height) {
+  const { client, pathComponents } = extractClientAndPathComponents(rawSrc);
+  const compactedParams = compactParamKeys(params, width, height);
+  return client.buildSrcSet(pathComponents.join("/"), compactedParams, options);
+}
+
+function warnInvalidAspectRatio(aspectRatio, config) {
+  const NODE_ENV = process.env.NODE_ENV;
+  const showARWarning =
+    aspectRatio != "" && aspectRatioIsValid(aspectRatio) === false;
+  if (
+    NODE_ENV !== "production" &&
+    showARWarning &&
+    config.warnings.invalidARFormat
+  ) {
+    console.warn(
+      `[Imgix] The aspect ratio passed ("${aspectRatio}") is not in the correct format. The correct format is "W:H".`
+    );
+  }
+}
+
+/**
+ * Validates that an aspect ratio is in the format w:h. If false is returned, the aspect ratio is in the wrong format.
+ */
+function aspectRatioIsValid(aspectRatio) {
+  if (typeof aspectRatio !== "string") {
+    return false;
+  }
+
+  return /^\d+(\.\d+)?:\d+(\.\d+)?$/.test(aspectRatio);
+}
+
+function buildChildProps(obj, src, attributeConfig, width, height, refType) {
+  const childProps = {
+    ...obj.props.htmlAttributes,
+    [attributeConfig.sizes]: obj.props.sizes,
+    className: obj.props.className,
+    width: width <= 1 ? null : width,
+    height: height <= 1 ? null : height,
+    [attributeConfig.src]: src,
```

Comment:
This shouldn't set `src` for `<source>` elements

Reply:
Oof, yes, good catch. Thanks!